src/v/v8_engine/executor.h

```diff
@@ -0,0 +1,232 @@
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "seastarx.h"
+#include "utils/concepts-enabled.h"
+#include "vassert.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/condition-variable.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/reactor.hh>
+#include <seastar/core/semaphore.hh>
+#include <seastar/core/sharded.hh>
+#include <seastar/core/smp.hh>
+
+#include <boost/lockfree/queue.hpp>
+
+#include <atomic>
+#include <chrono>
+#include <condition_variable>
+#include <cstddef>
+#include <list>
+#include <memory>
+#include <mutex>
+#include <semaphore.h>
+#include <v8.h>
+
+namespace v8_engine {
+
+namespace internal {
+
+// This class is used for notify seastar thread when executor thread will
+// complete task
+class condition {
+public:
+    explicit condition() noexcept;
+
+    // Run this function in seastar thread.
+    ss::future<> wait();
+
+    // Run this function in executor thread
+    void notify();
+
+private:
+    ss::file_desc _file_desc;
+    int _fd;
+    ss::pollable_fd _pollable_fd;
+    eventfd_t _event{0};
+};
+
+// Abstract class for executor task
+struct work_item {
+    virtual ~work_item() {}
+    virtual void process() = 0;
+    virtual void cancel() = 0;
+    virtual void on_end() = 0;
+    virtual std::chrono::milliseconds get_timeout() = 0;
+};
+
+// This class implement task for executor. Contains promise logic
+template<typename Func>
+struct task final : work_item {
+    task(Func&& f, std::chrono::milliseconds timeout)
+      : work_item()
+      , _func(std::forward<Func>(f))
+      , _timeout(timeout) {}
+
+    // Process task from executor in executor thread and set state in ss::future
+    void process() override {
+        try {
+            _func();
+            _state.set(true);
+        } catch (...) {
+            _state.set_exception(std::current_exception());
+        }
+        _on_done.notify();
+    }
+
+    void cancel() override { _func.cancel(); }
+
+    void on_end() override { _func.on_end(); }
+
+    std::chrono::milliseconds get_timeout() override { return _timeout; }
+
+    // Get future for seastar thread for wating when executor thread will
+    // complete task
+    ss::future<> get_future() {
+        return _on_done.wait().then([this] {
+            if (_state.failed()) {
+                return ss::make_exception_future<>(
+                  std::move(_state).get_exception());
+            } else {
+                return ss::make_ready_future<>();
+            }
+        });
+    }
+
+    Func _func;
+    ss::future_state<bool> _state;
+    condition _on_done;
+
+    std::chrono::milliseconds _timeout{};
+};
+
+class blocking_queue {
```

Comment:
I guess that this class could be renamed to something like `sharded_blocking_queue` since any shard can push.
Also, it's worth mentioning that alien thread can safely call pop.

Reply:
After @mmaslankaprv review 

Only executor will be sharded. We nedd to use invoke_on(0, ...) for submit task to executor

Reply:
I think that `blocking_queue` interface should be have no `signal()` method. Signalling should be done internally. There is enough to have two methods `put()` and `pop()`. `put()` should block when queue is full while `pop()` when empty

Reply:
yes, I lost this moment

Reply:
Done